README.md
# getrusage.js

[![NPM version](https://badgen.net/npm/v/getrusage.js?color=green)](https://www.npmjs.com/package/getrusage.js)

POSIX system function [`getrusage()`](https://www.gnu.org/software/libc/manual/html_node/Resource-Usage.html) is usually used to examine the resource usage of a process.

[`getrusage.js`](https://www.npmjs.com/package/getrusage.js) is a binding to Node.js via [Node-API](https://nodejs.org/api/n-api.html), and makes pre-build binaries for distribution, with **0 other dependencies!**

You don't need to compile (`node-gyp`) or download (`node-pre-gyp`) anything by postinstall.

## Support platforms

| Platform \ Nodejs                                  | node 12 | node 14 | node 16 | node 18 |
| -------------------------------------------------- | ------- | ------- | ------- | ------- |
| **Linux x64** <br>*<small>(gcc >= 4.9)</small>*    | ✓       | ✓       | ✓       | ✓       | 
| **Linux arm64** <br>*<small>(gcc >= 5.1)</small>*  | ✓       | ✓       | ✓       | ✓       | 
| **Linux arm32** <br>*<small>(gcc >= 5.1)</small>*  | ✓       | ✓       | ✓       | ✓       | 
| **macOS x64**                                      | ✓       | ✓       | ✓       | ✓       | 
| **macOS arm64** <br>*<small>(M1/M2 chips)</small>* | ✓       | ✓       | ✓       | ✓       | 

> Windows? not support yet, due to `getrusage()` is POSIX only.

## Install

```bash
npm i getrusage.js
```

## Usage

```ts
import {
  getrusage,
  RUSAGE_SELF,
  RUSAGE_CHILDREN,
} from 'getrusage.js'
import type { RUsage } from 'getrusage.js'

/**
 * `getrusage()` is equivalent to `getrusage(RUSAGE_SELF)`
 * `getrusage(RUSAGE_CHILDREN)` means usage of "All child processes (direct and indirect) that have already terminated"
 */
const usage: RUsage = getrusage()

console.log(usage)

// {
//   utime: 0.513607,
//   stime: 0.011146,
//   maxrss: 65421,
//   idrss: 0,
//   ixrss: 0,
//   isrss: 0,
//   minflt: 9753,
//   majflt: 12,
//   nswap: 0,
//   inblock: 0,
//   oublock: 0,
//   msgsnd: 0,
//   msgrcv: 0,
//   nsignals: 0,
//   nvcsw: 1347,
//   nivcsw: 10
// }
```

interface of `RUsage` and explain fields see:

```ts
/**
 * Resource Usage data struct
 * https://www.gnu.org/software/libc/manual/html_node/Resource-Usage.html#index-struct-rusage
 */
interface RUsage {
  /**
   * time spent in executing user instructions (seconds)
   * sum of tv_sec and tv_usec (unit: seconds)
   */
  utime: number;

  /**
   * time spent in operating system code on behalf of processes (seconds)
   * sum of tv_sec and tv_usec (unit: seconds)
   */
  stime: number;

  /**
   * the maximum of physical memory that processes used simultaneously (kilobytes)
   * always normalized unit in kilobytes on macOS
   */
  maxrss: number;

  /** deprecated fields in gnu: idrss / ixrss / isrss */
  idrss: number;
  ixrss: number;
  isrss: number;

  /** the number of page faults which were serviced without requiring any I/O (count) */
  minflt: number;

  /** the number of page faults which were serviced by doing I/O (count) */
  majflt: number;

  nswap: number;
  inblock: number;
  oublock: number;

  /** number of IPC messages sent (count) */
  msgsnd: number;

  /** number of IPC messages received (count) */
  msgrcv: number;

  /** number of signals received (count) */
  nsignals: number;

  /**
   * the number of times processes voluntarily invoked a context switch (count) 
   * (usually to wait for some service)
   */
  nvcsw: number;

  /**
   * the number of times an involuntary context switch took place (count)
   * because a time slice expired, or another process of higher priority was scheduled
   */
  nivcsw: number;
}
```

## License

MIT


CHANGELOG.md
# Changelog

## 0.2.7

### Patch Changes

- 9a213cf: make linux-x64 building file backwards-compatible to gcc-4.9

https://www.npmjs.com/package/getrusage.js

## [0.2.7](https://github.com/zthxxx/getrusage.js/compare/v0.2.1...v0.2.7)

### chore

- make `linux x64` building file backwards-compatible to `gcc-4.9`

### ci

- use install `gcc-4.9` and use `gcc-4.9` to building for `linux x64`

## [0.2.1](https://github.com/zthxxx/getrusage.js/compare/v0.1.2...v0.2.1)

### feature

- support `linux`/`macOS` x `x64`/`arm` by cross-platform compile on GitHub Actions; (`gcc >= 5.1`)

## [0.1.2](https://github.com/zthxxx/getrusage.js/tree/v0.1.2)

first publish version, support only `x64` in POSIX with `gcc >= 4.2`


tests/test.js
const fs = require('fs')
const path = require('path')
const util = require('util')
const assert = require('assert')
const { performance } = require('perf_hooks')

const core = require('@actions/core')
const promiseSpawn = require('@npmcli/promise-spawn')
const {
  getrusage,
  RUSAGE_SELF,
  RUSAGE_CHILDREN,
} = require('../lib')

const inspect = (obj) => util.inspect(obj, { colors: true, compact: false })

const assertStruct = (usage) => {
  const keys = new Set([
    'utime',
    'stime',
    'maxrss',
    'idrss',
    'ixrss',
    'isrss',
    'minflt',
    'majflt',
    'nswap',
    'inblock',
    'oublock',
    'msgsnd',
    'msgrcv',
    'nsignals',
    'nvcsw',
    'nivcsw',
  ])

  assert(
    Object.keys(usage)
      .filter((key) => !keys.has(key))
      .length === 0,
  )

  assert(
    Object.values(usage)
      .every((value) => (typeof value === 'number' && isFinite(value)))
  )
}

core.info(`
  Platform: ${process.platform}-${process.arch}
`)

core.startGroup(`RUSAGE flags`)
core.info(`RUSAGE_SELF: ${RUSAGE_SELF}`)
core.info(`RUSAGE_CHILDREN: ${RUSAGE_CHILDREN}`)
core.endGroup()



const stdioConfig = {
  stdioString: true,
  stdio: 'inherit',
}

const selfInitUsage = getrusage()

const pipeline = [
  () => {
    core.startGroup(`[getrusage] - self initial`)
    core.info(`selfInitUsage ${inspect(selfInitUsage)}`)
    core.endGroup()

    assertStruct(selfInitUsage)
    assert.ok(selfInitUsage.utime + selfInitUsage.stime > 0)
    assert.ok(selfInitUsage.maxrss > 0)
  },

  () => {
    const nonChildUsage = getrusage(RUSAGE_CHILDREN)

    core.startGroup(`[getrusage] - child none`)
    core.info(`nonChildUsage ${inspect(nonChildUsage)}`)
    core.endGroup()

    assertStruct(nonChildUsage)
    assert.equal(nonChildUsage.utime, 0)
    assert.equal(nonChildUsage.stime, 0)
    assert.equal(nonChildUsage.maxrss, 0)
  },

  () => promiseSpawn(
    'npm',
    ['run'],
    stdioConfig,
  ),
  
  () => promiseSpawn(
    'du',
    ['-sh', 'node_modules'],
    stdioConfig,
  ),

  () => {
    const childUsage = getrusage(RUSAGE_CHILDREN)

    core.startGroup(`[getrusage] - child done`)
    core.info(`childUsage ${inspect(childUsage)}`)
    core.endGroup()

    assertStruct(childUsage)
    assert.ok(childUsage.utime > 0)
    assert.ok(childUsage.stime > 0)
    assert.ok(childUsage.maxrss > 0)
  },

  () => {
    /**
     * allocate 50M heap
     * each element size of BigInt64Array is 8 bytes
     * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt64Array
     */
    const data = new BigInt64Array(50 * 1024 * 1024 / 8).fill(128n)

    if (!fs.existsSync('temp')) {
      fs.mkdirSync('temp')
    }
    fs.writeFileSync(path.join('temp', 'data.log'), data, { flag: 'w' })
  },

  () => {
    const selfUsage = getrusage()

    core.startGroup(`[getrusage] - self allocate 50M`)
    core.info(`selfUsage ${inspect(selfUsage)}`)
    core.endGroup()

    assertStruct(selfUsage)
    assert.ok(selfUsage.utime > 0)
    assert.ok(selfUsage.stime > 0)
    assert.ok(selfUsage.maxrss - selfInitUsage.maxrss > 50 * 1024)
  },

  () => {
    let piece = 2500
    while (piece > 0) {
      const now = performance.now()
      while (performance.now() - now < 1) {}
      piece -= 1
    }
  },

  () => {
    const selfUsage = getrusage(RUSAGE_SELF)

    core.startGroup(`[getrusage] - self: loop time`)
    core.info(`selfUsage ${inspect(selfUsage)}`)
    core.endGroup()

    assertStruct(selfUsage)
    assert.ok(selfUsage.utime > 2)
    assert.ok(selfUsage.stime > 0)
    assert.ok(selfUsage.maxrss > 0)
  },
]

const runTest = async () => {
  for(const step of pipeline) {
    await step()
  }
}

runTest()
  .catch((err) => {
    console.error(err)
    process.exit(1)
  })


src/index.ts
import path from 'path'
import { existsSync as isExists } from 'fs'


const platformArch = `${process.platform}-${process.arch}`

if (!isExists(path.join(__dirname, '..', 'build', `getrusage-${platformArch}.node`))) {
  throw Error(`
  getrusage.js not support for platform ${platformArch} now!
  `)
}

const binding = require(path.join('..', 'build', `getrusage-${platformArch}`))

export type RUSAGE_SELF = number
export type RUSAGE_CHILDREN = number
export const RUSAGE_SELF: RUSAGE_SELF = binding.RUSAGE_SELF
export const RUSAGE_CHILDREN: RUSAGE_CHILDREN = binding.RUSAGE_CHILDREN

/**
 * Resource Usage data struct
 * https://www.gnu.org/software/libc/manual/html_node/Resource-Usage.html#index-struct-rusage
 */
export interface RUsage {
  /**
   * time spent in executing user instructions (seconds)
   * sum of tv_sec and tv_usec (unit: seconds)
   */
  utime: number;
  /**
   * time spent in operating system code on behalf of processes (seconds)
   * sum of tv_sec and tv_usec (unit: seconds)
   */
  stime: number;
  /**
   * the maximum of physical memory that processes used simultaneously (kilobytes)
   * always normalized unit in kilobytes on macOS
   */
  maxrss: number;
  idrss: number;
  ixrss: number;
  isrss: number;
  /** the number of page faults which were serviced without requiring any I/O */
  minflt: number;
  /** the number of page faults which were serviced by doing I/O (count) */
  majflt: number;
  nswap: number;
  inblock: number;
  oublock: number;
  /** number of IPC messages sent */
  msgsnd: number;
  /** number of IPC messages received */
  msgrcv: number;
  /** number of signals received */
  nsignals: number;
  /**
   * the number of times processes voluntarily invoked a context switch
   * (usually to wait for some service)
   */
  nvcsw: number;
  /**
   * the number of times an involuntary context switch took place (count)
   * because a time slice expired, or another process of higher priority was scheduled
   */
  nivcsw: number;
}

export const getrusage = (who?: RUSAGE_SELF | RUSAGE_CHILDREN): RUsage => {
  const usage: RUsage = binding.getrusage(who)
  usage.utime = Number(usage.utime.toFixed(6))
  usage.stime = Number(usage.stime.toFixed(6))

  /**
   * due to macOS use BSD getrusage which unit is bytes,
   * while linux use GNU getrusage and the unit is kilobytes
   */
  if (process.platform === 'darwin') {
    usage.maxrss = Math.round(usage.maxrss / 1024)
  }

  return usage
}



getrusage/getrusage.cc
/**
 * modify from https://github.com/andrasq/node-qrusage/blob/1.5.2/qrusage.cc
 */

#include <nan.h>
using v8::FunctionTemplate;

#include <sys/time.h>
#include <sys/resource.h>

// extract the integer passed as the first argument
static int getIntArg( Nan::NAN_METHOD_ARGS_TYPE info, int index ) {
    int defaultValue = RUSAGE_SELF;
    return Nan::To<int32_t>(info[index]).FromMaybe(defaultValue);
}

NAN_METHOD( getrusage_array ) {
    struct rusage ru;
    int who = (info[0]->IsNumber()) ? getIntArg(info, 0) : RUSAGE_SELF;

    getrusage(who, &ru);

    // faster to zero-detect than to create new Number every time
    v8::Local<v8::Number> zero = Nan::New(0);
    #define _number(v)     (((v) > 0) ? Nan::New((double)v) : zero)

    // note: nan-2.2.0 is unable to Nan::New() a long (ambiguous), cast to double
    v8::Local<v8::Array> usage_array = Nan::New<v8::Array>(16);
    Nan::Set(usage_array, 0, Nan::New((double)ru.ru_utime.tv_sec + (double)ru.ru_utime.tv_usec * 1e-6));
    Nan::Set(usage_array, 1, Nan::New((double)ru.ru_stime.tv_sec + (double)ru.ru_stime.tv_usec * 1e-6));
    Nan::Set(usage_array, 2, Nan::New((double)ru.ru_maxrss));
    Nan::Set(usage_array, 4,  _number(ru.ru_ixrss));
    Nan::Set(usage_array, 3,  _number(ru.ru_idrss));
    Nan::Set(usage_array, 5,  _number(ru.ru_isrss));
    Nan::Set(usage_array, 6, Nan::New((double)ru.ru_minflt));
    Nan::Set(usage_array, 7,  _number(ru.ru_majflt));
    Nan::Set(usage_array, 8,  _number(ru.ru_nswap));
    Nan::Set(usage_array, 9,  _number(ru.ru_inblock));
    Nan::Set(usage_array, 10, _number(ru.ru_oublock));
    Nan::Set(usage_array, 11, _number(ru.ru_msgsnd));
    Nan::Set(usage_array, 12, _number(ru.ru_msgrcv));
    Nan::Set(usage_array, 13, _number(ru.ru_nsignals));
    Nan::Set(usage_array, 14, Nan::New((double)ru.ru_nvcsw));
    Nan::Set(usage_array, 15, Nan::New((double)ru.ru_nivcsw));

    info.GetReturnValue().Set(usage_array);
}

NAN_MODULE_INIT(Init) {
    Nan::Set(target, Nan::New("getrusageArray").ToLocalChecked(),       Nan::GetFunction(Nan::New<FunctionTemplate>(getrusage_array)).ToLocalChecked()),
    Nan::Set(target, Nan::New("RUSAGE_SELF").ToLocalChecked(),          Nan::New(RUSAGE_SELF)),
    Nan::Set(target, Nan::New("RUSAGE_CHILDREN").ToLocalChecked(),      Nan::New(RUSAGE_CHILDREN)),
    ((void)0) ;
}

NODE_MODULE(getrusage, Init)
